### ISSUE TITLE
[DISCUSSION] Can MAPL currently support multiple domain decompositions?

### ISSUE THREAD

**Issue by yantosca** [geoschem/GCHP#106]
Hi all.  Mike Long (@msl3v) and I are working on consolidating the chemistry mechanisms out of the old clunky GEOS-Chem modules (like sulfate_mod.F90 and gckpp_HetRates.F90) directly into KPP.  The question of optimizing domain decompositions for chemistry came up.

We understand that right now that the domain decomposition for the FV3 dycore is used throughout a GCHP simulation for all operations.  But we were wondering if it is possible to have multiple domain decompositions (e.g. one for advection, and a different one for chemistry that would minimize having too many sunrise/sunset boxes in one core).

Also, has anyone tested what the overhead of adding a second domain decomposition would be?  As long as the overhead isn't greater than the benefit of optimizing chemistry with a different domain decomposition, we would be interested in pursuing it.

Anyway, we weren't sure what is currently possible with GCHP via MAPL so would appreciate hearing your thoughts & comments.

@lizziel @LiamBindle @sdeastham @tclune @mathomp4 @msl3v @christophkeller @WilliamDowns

**Comment by sdeastham** [comment]
Hi Bob,

I think the problem with using multiple decompositions is that you'd have to shuttle the information around for every time step. A similar question would be how non-optimal it would be to have a "chemistry-optimized" rather than "transport-optimized" decomposition - to my knowledge, CESM (for example) actively chooses to give each core two domains which are antipodes of each other. That would in theory solve the problem. Could we do the same here? i.e. if we have N cores, split up the domain into 2*N subdomains and give each core the opposing faces so that the number of lit/unlit boxes are matched? It wouldn't be perfect (the cores with the terminator would still be worst) but would be better, without making communication for advection a total nightmare.

I do think that MAPL has other, more flexible options, but this seems like it would be somewhat straightforward?

**Comment by tclune** [comment]
Going back and forth between two layouts is very expensive.  Lots of data must be transferred.   It is possible that the communication overhead is worthwhile.   E.g.,  GEOS does this for SW radiation for a rather significant performance improvement.  (Not quite 2x.)   My estimates with GMI ~10 years ago were that chemistry would not have nearly so favorable of a benefit from moving back around.

MAPL itself generally assumes that all the components are on the same grid+layout, but, e.g., ocean and atmosphere are not, so it is possible to add components that would use a different layout.  The parent component would be responsible for the necessary redistribution.    Unlike the ocean though, the optimal balance for chemistry presumably changes every timestep due to the location of light/dark, so the better analogy is with SW radiation.     It would be a nontrivial, but quite feasbile, project to measure this with GCHP.

**Comment by LiamBindle** [comment]
From my recent timing tests, it seems like even a moderate improvment to the balancing would cause a pretty significant speed up:
![image](https://user-images.githubusercontent.com/12772306/121204453-220c0600-c834-11eb-8c36-08d554c0c4ff.png)

(Note: "waiting for chemistry" here is the exlusive time for the top-level `--Run` step. I've assumed it's a barrier waiting for chemistry to finish)


@sdeastham, do you know how straightforward that antipodal decomposition would be to implement? Seems like a good solution to me.

**Comment by tclune** [comment]
@LiamBindle it is hard to read from these figures.  Can you compute the ratio of the slowest core (doing chem kernel) to the average core for any/all of the cases?    That is the theoretical improvement to chem performance if communication overhead were 0.   With GMI, we found that the ratio was about 2x, and that a simplistic model to predict the most expensive locations in advance reduced the effective ration to something a bit lower.   (And there is overhead even computing the new layout for arbitrary decompositions - but this should be less than the raw transfer of data.)

**Comment by sdeastham** [comment]
@LiamBindle no idea - but hopefully not too difficult, as it we would use the existing decomposition but then oversubscribe cores with 2 domains each. I _think_ the only tricky step would be making sure that each core got the antipodal domains. @tclune do you have any idea where this decomposition approach lies on the spectrum from easy to insane?

**Comment by tclune** [comment]
The antipodal approach involves relatively little code change if done cleverly.   Basically one just constructs a different MPI comm in the CAP layer that puts things where you want them.    It  takes a nontrivial amount of caffeine to get that shuffle straight in ones head, but it should only be 10-20 lines of code.   GEOS itself is blissfully unaware.

We've done this before to try to just compact the layout for FV itself.    Currently the default layout goes as follows.   Suppose we have 24x24 cores for a face of the cube, and each node has 40 cores.     What will happen is the 24 cores on the lowest edge of the face will be assigned to the first node,  as will the first 16 in the next row up from the edge.   The remaining 8 will be assigned to the 2nd node.  (Picture would be easy to understand, sorry.)        This results in each node having a rather poor aspect ratio with high surface area to volume which is suboptimal for communication off node.  (On node communication is faster.)    We introduced a communicator to change this to something much more compact, but found the improvement was negligible.    I wish now though we had tried a completely random shuffle - just to be sure that there was some effect and that we had not missed something in our quick-and-dirty attempt.

Basically someone just needs to define a map  from  r ---> r'  where r is the default rank and r' is the rank in the new communicator for all cores.   I can help if someone wants to do the leg work.  Leaving on a 1 week vacation this Thursday though.

**Comment by LiamBindle** [comment]
Here are the min, avg, max times for the GCHP comp for the C180 on 576 core test from the figures above:
```
  Times for GCHP
TOTAL                   :    9893.823   13103.567   16306.995
RUN                     :    9770.605   12980.360   16183.784
GCHPctmEnv              :     168.905     172.360     178.384
GCHPchem                :    6600.656    9809.476   13015.178
DYNAMICS                :    3117.227    3121.372    3124.892
```

If I understand how to read these numbers correctly, the slowest DE's wall time for GCHPchem was 13015.178 sec and the average was 9809.476. So only 1.3x?

**Comment by tclune** [comment]
Yes - that would be my understanding.     The GMI scientists were comparing max to min and thinking that there was more opportunity.  An understandable misconception.  (Of course that was 7x for GMI at the time!)
